treeherder/perfalert/perfalert/__init__.py

```diff
@@ -83,29 +83,37 @@ def __init__(self, push_timestamp, value, testrun_timestamp=None,
         self.state = state
 
     def __cmp__(self, o):
+        # only compare value to make sorting deterministic
+        # in cases where we have multiple datapoints with
+        # the same push timestamp
         return cmp(
-                (self.push_timestamp, self.testrun_timestamp),
-                (o.push_timestamp, o.testrun_timestamp),
-                )
+            (self.push_timestamp, self.testrun_timestamp, self.value),
+            (o.push_timestamp, o.testrun_timestamp, o.value),
+        )
 
     def __eq__(self, o):
+        # same as for __cmp__: only compare value to make
+        # comparisons determistic
         return cmp(
-                (self.testrun_timestamp, self.value, self.buildid, self.machine_id),
-                (o.testrun_timestamp, o.value, o.buildid, o.machine_id),
-                ) == 0
+            (self.testrun_timestamp, self.value, self.buildid),
+            (o.testrun_timestamp, o.value, o.buildid),
+        ) == 0
 
     def __ne__(self, o):
         return not self == o
 
     def __repr__(self):
-        return "<%s: %.3f, %i, %s>" % (self.buildid, self.value,
-                                       self.testrun_timestamp, self.machine_id)
+        return "<%s: %.3f, %i, %i, %s>" % (self.buildid, self.value,
+                                           self.testrun_timestamp,
+                                           self.testrun_id,
+                                           self.state)
 
     def __str__(self):
-        return "Build %s on %s %s %s %s" % (self.buildid,
-                                            self.testrun_timestamp,
-                                            self.push_timestamp, self.value,
-                                            self.machine_id)
+        return "Build %s on %s %s %s %.3f" % (self.buildid,
+                                              self.testrun_timestamp,
+                                              self.push_timestamp,
+                                              self.testrun_id,
+                                              self.value)
```

Comment:
Yeah this is crazy, we only need one of these methods. I fixed the perfalert code to be much simpler and consolidated this in the process.